qualtran/bloqs/mod_arithmetic/mod_addition.py

```diff
@@ -316,9 +360,38 @@ def signature(self) -> 'Signature':
             ]
         )
 
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', ctrl: 'Soquet', x: 'Soquet', y: 'Soquet'
+    ) -> Dict[str, 'SoquetT']:
+        if isinstance(self.bitsize, sympy.Expr):
+            raise DecomposeTypeError("Cannot decompose symbolic `bitsize`.")
+        x_split = bb.split(x)
+        for i in range(self.bitsize):
+            and_ctrl = [ctrl, x_split[i]]
+            and_ctrl, ancilla = bb.add(And(), ctrl=and_ctrl)
+            ancilla, y = bb.add(
+                CModAddK(
+                    k=((self.k * 2 ** (self.bitsize - 1 - i)) % self.mod),
+                    bitsize=self.bitsize,
+                    mod=self.mod,
+                ),
+                ctrl=ancilla,
+                x=y,
+            )
+            and_ctrl = bb.add(And().adjoint(), ctrl=and_ctrl, target=ancilla)
+            ctrl = and_ctrl[0]
+            x_split[i] = and_ctrl[1]
+        x = bb.join(x_split, dtype=QUInt(self.bitsize))
+
+        return {'ctrl': ctrl, 'x': x, 'y': y}
+
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> 'BloqCountDictT':
         k = ssa.new_symbol('k')
-        return {CModAddK(k=k, bitsize=self.bitsize, mod=self.mod): self.bitsize}
+        return {
+            CModAddK(k=k, bitsize=self.bitsize, mod=self.mod): self.bitsize,
+            And(): self.bitsize,
```

Comment:
this changes the cost of the bloq. Do you have a reference for the present construction? We at least need a note about the difference between gidney 2019